src/Data/Text/Utf8/AhoCorasick/Automaton.hs

```diff
@@ -374,67 +379,137 @@ uAt = UVector.unsafeIndex
+      -- Transition matched!
+      | otherwise = collectMatches offset remaining acc $ transitionState t
+      where !t = rootAsciiTransitions `uAt` fromIntegral cu
+
+    -- This function can't be inlined since it is self-recursive.
+    {-# NOINLINE lookupTransition #-}
+    lookupTransition :: Int -> Int -> a -> CodeUnit -> Word32 -> State -> Int -> a
+    lookupTransition !offset !remaining !acc !cu !cus !state !i
+      -- There is no transition for the given input. Follow the fallback edge,
+      -- and try again from that state, etc. If we are in the base state
+      -- already, then nothing matched, so move on to the next input.
+      | transitionIsWildcard t =
+        if state == initialState
+          then consumeInput offset remaining acc state
+          else followEdge offset remaining acc cu cus (transitionState t)
+      -- We found the transition, switch to that new state, possibly matching the rest of cus.
+      -- NOTE: This comes after wildcard checking, because the code unit of
+      -- the wildcard transition is 0, which is a valid input.
+      | transitionCodeUnit t == cu =
+        if cus == 0
+          then collectMatches offset remaining acc (transitionState t)
+          else followCodeUnits offset remaining acc cus (transitionState t)
+      -- The transition we inspected is not for the current input, and it is not
+      -- a wildcard either; look at the next transition then.
+      | otherwise =
+        lookupTransition offset remaining acc cu cus state $ i + 1
+
+      where
+        !t = transitions `uAt` i
+
+    -- TODO: In this case, we could avoid the loop in followCodeUnits and just follow a single edge.
+    -- Should be worth investigating but could lead to some code duplication.
+    {-# INLINE followCodeUnit #-}
+    followCodeUnit :: Int -> Int -> a -> CodeUnit -> State -> a
+    followCodeUnit !offset !remaining !acc !cu0 !state =
+      followCodeUnits offset remaining acc (fromIntegral cu0) state
+
+    {-# INLINE follow2CodeUnits #-}
+    follow2CodeUnits :: Int -> Int -> a -> CodeUnit -> CodeUnit -> State -> a
+    follow2CodeUnits !offset !remaining !acc !cu0 !cu1 !state =
+      followCodeUnits offset remaining acc (fromIntegral cu0 .|. fromIntegral cu1 `shiftL` 8) state
+
+    {-# INLINE follow3CodeUnits #-}
+    follow3CodeUnits :: Int -> Int -> a -> CodeUnit -> CodeUnit -> CodeUnit -> State -> a
+    follow3CodeUnits !offset !remaining !acc !cu0 !cu1 !cu2 !state =
+      followCodeUnits offset remaining acc (fromIntegral cu0 .|. fromIntegral cu1 `shiftL` 8 .|. fromIntegral cu2 `shiftL` 16) state
+
+    {-# INLINE follow4CodeUnits #-}
+    follow4CodeUnits :: Int -> Int -> a -> CodeUnit -> CodeUnit -> CodeUnit -> CodeUnit -> State -> a
+    follow4CodeUnits !offset !remaining !acc !cu0 !cu1 !cu2 !cu3 !state =
+      followCodeUnits offset remaining acc (fromIntegral cu0 .|. fromIntegral cu1 `shiftL` 8 .|. fromIntegral cu2 `shiftL` 16 .|. fromIntegral cu3 `shiftL` 24) state
```

Comment:
This packing/unpacking could be part of the reason why it's slow. I wonder if we could simplify this some more.